### ISSUE TITLE
--diff should show the corresponding applied rules

### ISSUE THREAD

**Issue by vvo** [millermedeiros/esformatter#331]
What I find hard to do is tuning esformatter for some rules. Because when doing `esformatter --diff`, we do not know what would be the the applied rules.

Do you think that's feasible?

**Comment by millermedeiros** [comment]
it wasn't really clear to me how that feature would work, can you provide the expected output? (even if it's just a rough idea, just so we can get the conversation going)

**Comment by vvo** [comment]
Given this error:

![2015-07-08-154844_1052x119_scrot](https://cloud.githubusercontent.com/assets/123822/8571761/deba2976-2588-11e5-859f-6f18d614a92a.png)

We could have:

```sh
77 | <LF>
--  | rule: lineBreak.after.expre..: >= 1
78 | zuulConfig.browsers =<LF>
79 |   process.env.TRAVIS_PULL_REQUEST && process.env.TRAVIS_PULL_REQUEST !== 'false' ?<LF>
80 |     browsers.pullRequest :<LF>
81 |     browsers.all;<LF>
```

?

**Comment by millermedeiros** [comment]
that is tricky to implement.. right now we just compare the input with the output and do a [basic string diff](https://github.com/millermedeiros/disparity).

to know which parts of the code are not obeying the rules we would need to store the differences on all the line break and white space `limit` calls (eg. expected 2 line breaks, found 1) and after the whole transformation it would need to map the diffs with the line number (code formatting starts from the "leaf" nodes, so line numbers change a lot during the transformation).

indentation is a totally different process and we don't really know how many indents the original program had.. (and there is no precise way to do it because of mixed indentation and odd numbers..)

but I agree that it would be easier to configure esformatter with some sort of help, maybe we could do it through an UI that allow user to hover the code fragments and display the token/node type and which settings are applied to that part of the code..

**Comment by vvo** [comment]
Yep esformatter is great but was hard to configure, once set it's good. Good luck, I was only raising the issue, I really can't help more :)

**Comment by millermedeiros** [comment]
@vvo keep the bug reports and suggestions coming. it's good to have feedback.

**Comment by hadri3n** [comment]
+1
